### ISSUE TITLE
System.Runtime.CompilerServices.Unsafe causing issues with ASP.NET CORE app 2.1

### ISSUE THREAD

**Issue by ajayban** [dotnet/corefx#36182]
Hello All, 

One of customer I was working with reported this assembly load failure issue with his .NET Core 2.1 app. Below is the call stack:
```
System.IO.FileLoadException: Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
File name: 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.StringUtilities.TryGetAsciiString(Byte* input, Char* output, Int32 count)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.HttpUtilities.GetAsciiStringNonNullCharacters(Span`1 span)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.OnOriginFormTarget(HttpMethod method, HttpVersion version, Span`1 target, Span`1 path, Span`1 query, Span`1 customMethod, Boolean pathEncoded)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.OnStartLine(HttpMethod method, HttpVersion version, Span`1 target, Span`1 path, Span`1 query, Span`1 customMethod, Boolean pathEncoded)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpParser`1.ParseRequestLine(TRequestHandler handler, Byte* data, Int32 length)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpParser`1.ParseRequestLine(TRequestHandler handler, ReadOnlySequence`1& buffer, SequencePosition& consumed, SequencePosition& examined)
```
We were able to resolve the issue by removing the runtime reference according to https://github.com/dotnet/corefx/issues/23988#issuecomment-329200879.

This looks like this issue has reappear: #23988. Can someone please look into this?

**Comment by ericstj** [comment]
Do you have a repro project, or barring that an msbuild binlog for the failing project?  Thanks!

**Comment by ajayban** [comment]
Hello,

I do not have repro files.

However, as per the customer - The issue was with version 2.1.0 of Microsoft.AspNetCore.All (where System.Runtime.CompilerServices.Unsafe is one of the included dll), and Microsoft.EntityFrameworkCore.SqlServer (which has some other dll with hard coded dependency).

Newer versions for both (2.1.8 and 2.1.4 respectively) did not exhibit such behavior.

I’m guessing the 2.1.0 batch has some difference in behavior in terms of dependency which got removed in the later versions.

**Comment by ericstj** [comment]
It sounds like an issue with conflict resolution, but without a repro we cannot triage this. Either a project file with the dependencies that exhibit this problem or an msbuild binlog should show what's going on.  Perhaps the deps file for the failing project would work too, but that might not be enough. Somehow the app is getting the old copy of S.R.CS.Unsafe in the output and is downgrading what the shared framework expects.  I can hypothesize here but we really need a repro to follow up.  @natemcmaster @dsplaisted 

I gave the following a try and it worked for me:
```
  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.All" Version="2.1.0" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="2.1.0" />
  </ItemGroup>
```

**Comment by vblz** [comment]
Have the same trouble. But it can be reproduced at only one host, at neighboring computer there are no troubles with same project. 
Logs:
```
Microsoft.AspNetCore.Server.Kestrel:Information: Connection id "0HLLK0ULQ3DJA" request processing ended abnormally.

System.IO.FileLoadException: Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
File name: 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.StringUtilities.TryGetAsciiString(Byte* input, Char* output, Int32 count)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.HttpUtilities.GetAsciiStringNonNullCharacters(Span`1 span)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.OnOriginFormTarget(HttpMethod method, HttpVersion version, Span`1 target, Span`1 path, Span`1 query, Span`1 customMethod, Boolean pathEncoded)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.OnStartLine(HttpMethod method, HttpVersion version, Span`1 target, Span`1 path, Span`1 query, Span`1 customMethod, Boolean pathEncoded)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpParser`1.ParseRequestLine(TRequestHandler handler, Byte* data, Int32 length)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpParser`1.ParseRequestLine(TRequestHandler handler, ReadOnlySequence`1& buffer, SequencePosition& consumed, SequencePosition& examined)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpParser`1.Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.IHttpParser<TRequestHandler>.ParseRequestLine(TRequestHandler handler, ReadOnlySequence`1& buffer, SequencePosition& consumed, SequencePosition& examined)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.TakeStartLine(ReadOnlySequence`1 buffer, SequencePosition& consumed, SequencePosition& examined)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.ParseRequest(ReadOnlySequence`1 buffer, SequencePosition& consumed, SequencePosition& examined)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.TryParseRequest(ReadResult result, Boolean& endConnection)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequestsAsync[TContext](IHttpApplication`1 application)


Microsoft.AspNetCore.Server.Kestrel:Debug: Connection id "0HLLK0ULQ3DJA" disconnecting.
Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets:Debug: Connection id "0HLLK0ULQ3DJA" sending FIN.
Microsoft.AspNetCore.Server.Kestrel:Debug: Connection id "0HLLK0ULQ3DJA" stopped.
```

Also `curl` says `curl: (52) Empty reply from server` for the request.

Version:  "Microsoft.AspNetCore.All (2.1.0)"

**Comment by ericstj** [comment]
Can you double check which version of the shared frameworks are installed on the failing server?

This can be done by running `dotnet --info`.

Can you also share the deps.json and runtimeconfig.json from the failing app?

@natemcmaster @wtgodbe I suspect this is due to a change in assembly version of Unsafe during servicing, and then running on an ASP.NET shared framework without that change.

**Comment by vblz** [comment]
This trouble was on PC, starting the app with visual studio. The bug was still after installing newest `dotnet sdk 2.2.105`, but without rebooting. A version of VS is 15.9.5.
The issue was solved by updating `Microsoft.AspNetCore.All` to 2.1.1. Today I tried to reproduce the issue and downgrade back to 2.1.0 and issue reproduces every request and app launch.

And the strange behavior: the issue is gone after remove dependency `Directum.AspNetCore.Utils.Jaeger`. After restoring the dependency the trouble is returned and reproducing.


[jsons.zip](https://github.com/dotnet/corefx/files/3027638/jsons.zip)

<details><summary>dotnet --info</summary>
     
     Пакет SDK для .NET Core (отражающий любой global.json):
     Version:   2.2.105
     Commit:    7cecb35b92

    Среда выполнения:
     OS Name:     Windows
     OS Version:  10.0.14393
     OS Platform: Windows
     RID:         win10-x64
     Base Path:   C:\Program Files\dotnet\sdk\2.2.105\

    Host (useful for support):
      Version: 2.2.3
      Commit:  6b8ad509b6

    .NET Core SDKs installed:
      2.0.3 [C:\Program Files\dotnet\sdk]
      2.1.1-preview-007165 [C:\Program Files\dotnet\sdk]
      2.1.2 [C:\Program Files\dotnet\sdk]
      2.1.4 [C:\Program Files\dotnet\sdk]
      2.1.200 [C:\Program Files\dotnet\sdk]
      2.1.201 [C:\Program Files\dotnet\sdk]
      2.1.202 [C:\Program Files\dotnet\sdk]
      2.1.300 [C:\Program Files\dotnet\sdk]
      2.1.302 [C:\Program Files\dotnet\sdk]
      2.1.400 [C:\Program Files\dotnet\sdk]
      2.1.401 [C:\Program Files\dotnet\sdk]
      2.1.403 [C:\Program Files\dotnet\sdk]
      2.1.500 [C:\Program Files\dotnet\sdk]
      2.1.502 [C:\Program Files\dotnet\sdk]
      2.1.503 [C:\Program Files\dotnet\sdk]
      2.2.105 [C:\Program Files\dotnet\sdk]

    .NET Core runtimes installed:
      Microsoft.AspNetCore.All 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
      Microsoft.AspNetCore.All 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
      Microsoft.AspNetCore.All 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
      Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
      Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
      Microsoft.AspNetCore.All 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
      Microsoft.AspNetCore.All 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
      Microsoft.AspNetCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
      Microsoft.AspNetCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
      Microsoft.AspNetCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
      Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
      Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
      Microsoft.AspNetCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
      Microsoft.AspNetCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
      Microsoft.NETCore.App 2.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      Microsoft.NETCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      Microsoft.NETCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      Microsoft.NETCore.App 2.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      Microsoft.NETCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      Microsoft.NETCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      Microsoft.NETCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

    To install additional .NET Core runtimes or SDKs:
      https://aka.ms/dotnet-download
</details>

**Comment by ericstj** [comment]
So it look like the app has a local copy of Unsafe.dll
```
      "System.Runtime.CompilerServices.Unsafe/4.5.1": {
        "runtime": {
          "lib/netcoreapp2.0/System.Runtime.CompilerServices.Unsafe.dll": {
            "assemblyVersion": "4.0.4.0",
            "fileVersion": "0.0.0.0"
          }
        },
        "compile": {
          "ref/netstandard2.0/System.Runtime.CompilerServices.Unsafe.dll": {}
        }
```

This would downgrade the runtime, but should only do so if the run-time was missing assembly-version information in its deps file and the runtime had dependencies to the higher version.  I poked around the different deps files and found ASP.NET was missing in Microsoft.AspNetCore.App\2.1.7 and carried the higher version of S.R.CS.Unsafe.  This was fixed in Microsoft.AspNetCore.App\2.1.8: in this shared framework the shared framework deps file contains the assembly version.  I believe that explains this behavior.  I noticed your failing machine is missing the 2.1.8 aspnet shared framework, if you update to 2.1.8 does it go away.  This is all pointing to an issue with the way that the ASP.NET shared frameworks are built.

If you have the machine in a state that's reproducing the issue, does installing the latest ASP.NET 2.1 runtime fix it?  https://dotnet.microsoft.com/download/dotnet-core/2.1

**Comment by vblz** [comment]
After installing `2.1.505` the trouble is not reproducing.

But just for information, why changing dependency `Microsoft.AspNetCore.All` to ` 2.1.1` also solves the issue on previous versions `2.1.503` and `2.2.105`?

**Comment by ericstj** [comment]
I believe that since ASP.NET isn't providing a platform manifest to the .NET SDK's conflict resolution it ends up relying on package-identity to determine what gets copied into the app or not.

The 2.1.1 version of Microsoft.AspNetCore.All contains System.Runtime.CompilerServices.Unsafe/4.5.1 in its closure, so the file is removed from the application.

The 2.1.0 version of Microsoft.AspNetCore.All contains contains only System.Runtime.CompilerServices.Unsafe/4.5.0, so a reference to 4.5.1 remains in the app.  Even though the application is targeting the 2.1.0 version of the shared framework it rolls forward to a serviced version (eg 2.1.7).  That serviced version needs a newer version of s.r.cs.us to satisfy its references.  So although 4.5.1 package version of this was an upgrade for the compiled shared framework, it's a downgrade in the serviced-runtime-shared framework.

The general solution for dealing with this is the host making the decision based on versions in the deps file, which it looks like ASP.NET started adding in 2.1.8.  It surprises me that it took this long, but it does appear to be fixed.  /cc @DamianEdwards @vitek-karas for awareness.

**Comment by ericstj** [comment]
Closing as this an issue in ASP.NET shared frameworks.  If you still feel there is an issue, please open in https://github.com/aspnet/AspNetCore/issues

**Comment by natemcmaster** [comment]
> The general solution for dealing with this is the host making the decision based on versions in the deps file, which it looks like ASP.NET started adding in 2.1.8. It surprises me that it took this long, but it does appear to be fixed.

This was the general solution. It was fixed in 2.1.5 by and update to the host, which was released like 6 or 7 months ago. We regressed in 2.1.7 on Windows only. See https://github.com/aspnet/AspNetCore/issues/3503
